File: go/oop-patterns/behavioral-patterns/visitor/visitor.go
package main

import "fmt"

// Каждый компонент, который нужно посетить должен содержать метод,
// который позволит принять посетителя
type component interface {
	accept(visitor visitor)
}

type concreteComponentA struct {
}

func (cc *concreteComponentA) accept(visitor visitor) {
	visitor.visitConcreteComponentA(cc)
}
func (cc *concreteComponentA) exclusiveMethodOfConcreteComponentA() string {
	return "A"
}

type concreteComponentB struct {
}

func (cc *concreteComponentB) accept(visitor visitor) {
	visitor.visitConcreteComponentB(cc)
}
func (cc *concreteComponentB) exclusiveMethodOfConcreteComponentB() string {
	return "B"
}

// Каждый посетитель должен иметь методы для посещения конкретных компонентов
type visitor interface {
	visitConcreteComponentA(element *concreteComponentA)
	visitConcreteComponentB(element *concreteComponentB)
}

type concreteVisitor1 struct {
}

func (cv *concreteVisitor1) visitConcreteComponentA(element *concreteComponentA) {

	fmt.Println(element.exclusiveMethodOfConcreteComponentA(), "+ ConcreteVisitor1")

}

func (cv *concreteVisitor1) visitConcreteComponentB(element *concreteComponentB) {

	fmt.Println(element.exclusiveMethodOfConcreteComponentB(), "+ ConcreteVisitor1")

}

type concreteVisitor2 struct {
}

func (cv *concreteVisitor2) visitConcreteComponentA(element *concreteComponentA) {

	fmt.Println(element.exclusiveMethodOfConcreteComponentA(), "+ ConcreteVisitor2")

}

func (cv *concreteVisitor2) visitConcreteComponentB(element *concreteComponentB) {

	fmt.Println(element.exclusiveMethodOfConcreteComponentB(), "+ ConcreteVisitor2")

}

//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
//------------------------------- Клиентский код -------------------------------
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func main() {
	comp1 := concreteComponentA{}
	comp2 := concreteComponentB{}

	fmt.Println(
		"The client code works with all visitors via the base Visitor interface:")
	visitor1 := &concreteVisitor1{}
	visitor2 := &concreteVisitor2{}
	fmt.Println("It allows the same client code to work with different types of visitors:")

	comp1.accept(visitor1)
	comp1.accept(visitor2)
	comp2.accept(visitor1)
	comp2.accept(visitor2)
}

File: typescript/oop-patterns/behavioral-patterns/visitor/visitor.ts
// Каждый компонент, который нужно посетить должен содержать метод,
// который позволит принять посетителя
interface Component {
  accept(visitor: Visitor): void;
}

class ConcreteComponentA implements Component {
  public accept(visitor: Visitor): void {
    visitor.visitConcreteComponentA(this);
  }
  public exclusiveMethodOfConcreteComponentA(): string {
    return 'A';
  }
}

class ConcreteComponentB implements Component {
  public accept(visitor: Visitor): void {
    visitor.visitConcreteComponentB(this);
  }

  public specialMethodOfConcreteComponentB(): string {
    return 'B';
  }
}

// Каждый посетитель должен иметь методы для посещения конкретных компонентов
interface Visitor {
  visitConcreteComponentA(element: ConcreteComponentA): void;
  visitConcreteComponentB(element: ConcreteComponentB): void;
}

class ConcreteVisitor1 implements Visitor {
  public visitConcreteComponentA(element: ConcreteComponentA): void {
    console.log(
      `${element.exclusiveMethodOfConcreteComponentA()} + ConcreteVisitor1`
    );
  }

  public visitConcreteComponentB(element: ConcreteComponentB): void {
    console.log(
      `${element.specialMethodOfConcreteComponentB()} + ConcreteVisitor1`
    );
  }
}

class ConcreteVisitor2 implements Visitor {
  public visitConcreteComponentA(element: ConcreteComponentA): void {
    console.log(
      `${element.exclusiveMethodOfConcreteComponentA()} + ConcreteVisitor2`
    );
  }

  public visitConcreteComponentB(element: ConcreteComponentB): void {
    console.log(
      `${element.specialMethodOfConcreteComponentB()} + ConcreteVisitor2`
    );
  }
}

//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
//------------------------------- Клиентский код -------------------------------
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

const comp1 = new ConcreteComponentA();
const comp2 = new ConcreteComponentB();

console.log(
  'Клиент работает с каждым посетителем через базовый интерфейс посетителя'
);
const visitor1 = new ConcreteVisitor1();
const visitor2 = new ConcreteVisitor2();
console.log('Что позволяет клиенту работать с различными посетителями:');

comp1.accept(visitor1);
comp1.accept(visitor2);
comp2.accept(visitor1);
comp2.accept(visitor2);
